python/tk_framework_desktopserver/shotgun/api_v2.py

```diff
@@ -621,6 +592,68 @@ def _get_lookup_hash(self, config_uri, project, entity_type):
             entity_type=entity_type
         )
 
+    def _get_pipeline_configuration_data(self, manager, project_entity, data):
+        """
+        Gathers all of the necessary data pertaining to the project's pipeline
+        configurations. This includes the PipelineConfiguration entity, the
+        contents and lookup hashes, and the associated descriptor object.
+
+        :param manager: A ToolkitManager.
+        :param dict project_entity: The Project entity dict.
+        :param dict data: The payload from the client.
+
+        :returns: A dictionary, keyed by PipelineConfiguration entity id, that
+            contains dictionaries with "contents_hash", "lookup_hash",
+            "descriptor", and "entity" keys.
+        :rtype: dict
+        """
+        if "config_data" in self.WSS_KEY_CACHE[self._wss_key]:
+            self.logger.debug("Pipeline config data found for %s", self._wss_key)
+        else:
+            config_data = dict()
+            pcs = self._get_pipeline_configurations(
+                manager,
+                project_entity,
+            ) or [dict(id=None, name="Primary")]
+
+            for pc in pcs:
+                self.logger.debug("Processing config: %s" % pc)
+
+                # The hash that acts as the key we'll use to look up our cached
+                # data will be based on the entity type and the pipeline config's
+                # descriptor uri. We can get the descriptor from the toolkit
+                # manager and pass that through along with the entity type from SG
+                # to the core hook that computes the hash.
+                manager.pipeline_configuration = pc["id"]
+
+                try:
+                    pc_descriptor = manager.resolve_descriptor(project_entity)
```

Comment:
more importantly, why do we have to do all these checks at runtime like this?

Could we have a more basic check that does a single site check to get the state of pcs and software and if that comes out clean then BAM we just load everything up from the cache - and if not, we kick into 'expensive mode' where we begin figuring out what needs recaching and then kick of the process to handle all that. Could the single cache check be for example `the number of pc records + latest pc updated date + number of sw records + latest sw updated date`?